### PR TITLE
Add external data utilities

### LINKED ISSUE #45: External data / index handlers

### PR DESCRIPTION
**Pull request by ZacharyEspiritu**
Resolves #45. 

Adds external data utilities for attaching engine-native objects to OpenSSL structs.

### CODE REVIEW

**Comment by ZacharyEspiritu** [comment]
I've marked this as ready for review because even though it technically depends on #70, it appears that GitHub will, at least, automatically update the base branch of the PR  to master once #70 is merged so then this can be merged immediately after.

File: src/bridge/ex_data_util/ex_data_util.cc
Comment (tonyliaoss) [inline src/bridge/ex_data_util/ex_data_util.cc]: Make this static if `rsa_index` is only used in this file?

File: src/bridge/ex_data_util/ex_data_util.cc
Comment (tonyliaoss) [inline src/bridge/ex_data_util/ex_data_util.cc]: Same here

File: src/bridge/ex_data_util/ex_data_util.cc
Comment (tonyliaoss) [inline src/bridge/ex_data_util/ex_data_util.cc:98]: What happens when `rsa_key` or `rsa` is a nullptr? Add a unit test for it?

File: src/bridge/ex_data_util/ex_data_util.cc
Comment (tonyliaoss) [inline src/bridge/ex_data_util/ex_data_util.cc:124]: Same here for `engine` and `data`

File: src/bridge/ex_data_util/ex_data_util.cc
Comment (tonyliaoss) [inline src/bridge/ex_data_util/ex_data_util.cc:110]: Same here

File: src/bridge/ex_data_util/ex_data_util.cc
Comment (tonyliaoss) [inline src/bridge/ex_data_util/ex_data_util.cc:136]: Same here

File: src/bridge/ex_data_util/ex_data_util.cc
Comment (ZacharyEspiritu) [reply src/bridge/ex_data_util/ex_data_util.cc]: Didn't mark as `static` since it's within an anonymous namespace, but if it's clearer can add `static`

File: src/bridge/ex_data_util/ex_data_util.cc
Comment (ZacharyEspiritu) [reply src/bridge/ex_data_util/ex_data_util.cc]: See above

File: src/bridge/ex_data_util/ex_data_util.cc
Comment (tonyliaoss) [reply src/bridge/ex_data_util/ex_data_util.cc]: Yeah I'd say mark them static too for clarity.

**Review by tonyliaoss — DISMISSED** [review]

File: src/bridge/ex_data_util/ex_data_util.h
Comment (ajfeldma-google) [inline src/bridge/ex_data_util/ex_data_util.h]: `s/Indicies/Indices`

File: src/bridge/ex_data_util/ex_data_util.cc
Comment (ajfeldma-google) [inline src/bridge/ex_data_util/ex_data_util.cc]: I think it would be better not to rely on the default constructor of `Status`. Is there an `OkStatus()` method of some sort that you can use?

File: src/bridge/ex_data_util/ex_data_util.cc
Comment (ajfeldma-google) [inline src/bridge/ex_data_util/ex_data_util.cc]: In general, prefer to spell out the return type explicitly unless it's really obvious: https://google.github.io/styleguide/cppguide.html#Type_deduction

In this case, knowing `ex_data` is `void*` is probably important.

File: src/bridge/ex_data_util/ex_data_util_test.cc
Comment (ajfeldma-google) [inline src/bridge/ex_data_util/ex_data_util_test.cc]: Similar to the comment above (and for the whole file), prefer to spell out the return type explicitly unless it's really obvious: https://google.github.io/styleguide/cppguide.html#Type_deduction

File: src/bridge/ex_data_util/ex_data_util_test.cc
Comment (ajfeldma-google) [inline src/bridge/ex_data_util/ex_data_util_test.cc:53]: If retrieving a null `RsaKey` is an error, should setting one be allowed?

**Review by ajfeldma-google — CHANGES_REQUESTED** [review]

File: src/bridge/ex_data_util/ex_data_util_test.cc
Comment (ZacharyEspiritu) [reply src/bridge/ex_data_util/ex_data_util_test.cc:53]: Setting to nullptr is useful since it allows us to null-out when we free the `RsaKey` at a later point. However, when you call `Get...`,  the caller should check if the returned `RsaKey` is null anyways (just in case the invariants got screwed up somewhere, or the application passed in a bad `RSA` struct). Thus, returning an error status on a null `RsaKey` reduces some repetition down the line for callers to have to do a `if (rsa_key == nullptr) { ... }` check every time there's a `Get...`.

File: src/bridge/ex_data_util/ex_data_util.cc
Comment (ajfeldma-google) [inline src/bridge/ex_data_util/ex_data_util.cc:107]: It looks like absl is going to have a free function `OkStatus()`: https://github.com/abseil/abseil-cpp/blob/1995c6a3c2f9080160d9d8716504dc004e5e1ec0/absl/status/status.h#L362

**Review by ajfeldma-google — APPROVED** [review]

**Review by tonyliaoss — APPROVED** [review]

**Pull request merged**
